(docs) - Add page for more detailed auth docs (#973)
FormidableLabs/urql: docs/advanced/authentication.md

    context: {
      ...operation.context,
      fetchOptions: {
        ...fetchOptions,
        headers: {
          ...fetchOptions.headers,
          "Authorization": authState.token,
        },
      },
    },
  };
}
```

First we check that we have an `authState` and a `token`. Then we apply it to the request `fetchOptions` as an `Authorization` header.
The header format can vary based on the API (e.g using `Bearer ${token}` instead of just `token`) which is why it'll be up to you to add the header
in the expected format for your API.

### Configuring `didAuthError`

This function lets the exchange know what is defined to be an API error for your API:

```js
const didAuthError = ({ error }) => {
  return error.graphQLErrors.some(
    e => e.extensions?.code === 'FORBIDDEN',
  );
}
```

For most graphQL APIs, the auth error is communicated as an error code, however it may also be a 401 HTTP response:

```js
const didAuthError = ({ error }) => {
  return error.graphQLErrors.some(
    e => error.response.status === 401,
  );
},
```

Then `didAuthError` returns `true`, it will trigger the exchange to trigger the logic for asking for reauthentication via `getAuth`.

> ### Configuring `getAuth` (after `didAuthError` returns `true`)

Reviewer: If "after `didAuthError` returns `true` is important information and there's no corresponding section that for instance says `false` instead, let's move this into the text below as a full explanation?

> `getAuth` will be triggered to refresh the authentication state if `didAuthError` returned `true` previously, due to at least one GraphQL request erroring out with an authentication error, as we defined previously.

Author: The two sections are "Configuring `getAuth` (fetch from storage)" and "Configuring `getAuth` (after `didAuthError` returns `true`)".

I could rephrase it to "after an auth error has happened"?

If your API doesn't support any sort of token refresh, this is where you should simply log the user out.

```js
const getAuth = async ({ authState }) => {
  if (!authState) {
    const token = localStorage.getItem('token');
    const refreshToken = localStorage.getItem('refreshToken');
    if (token && refreshToken) {
      return { token, refreshToken };
    }
    return null;
  }

  logout();

  return null;
}
```

> Here, `logout()` should clear your persisted storage of any tokens and set the `isLoggedIn` property as described above to `false`.
> 
> If you do have a token refresh endpoint, and a refresh token, you can attempt to get a new token for the user:

Reviewer: ```suggestion
Here, `logout()` is a placeholder that is called when we got an error, so that we can redirect to a login page again and clear our tokens from local storage or otherwise.

If we had a way to refresh our token using a refresh token, we can attempt to get a new token for the user first:
```

Just rewriting to keep the first-person plural consistent

```js
const getAuth = async ({ authState, mutate }) => {
  if (!authState) {
    const token = localStorage.getItem('token');
    const refreshToken = localStorage.getItem('refreshToken');
    if (token && refreshToken) {
      return { token, refreshToken };
    }
    return null;
  }

  const result = await mutate(refreshMutation, {
    token: authState!.refreshToken,
  });

  if (result.data?.refreshLogin) {
    localStorage.setItem('token', result.data.refreshLogin.token);
    localStorage.setItem('refreshToken', result.data.refreshLogin.refreshToken);

    return {
      token: result.data.refreshLogin.token,
      refreshToken: result.data.refreshLogin.refreshToken,
    };
  }

>   localStorage.clear();
>   logout();

Reviewer: ```suggestion
  // This is where auth has gone wrong and we need to clean up and redirect to a login page
  localStorage.clear();
  logout();
```

Just adding a comment here since this isn't reusable for the user's app

  return null;
}
```

> Here we use the special mutate function provided by the auth exchange to do the token refresh. If your auth is not handled via graphQL, you can

Reviewer: ```suggestion
Here we use the special mutate function provided by the auth exchange to do the token refresh. If your auth is not handled via GraphQL but a REST endpoint, you can
```

> use `fetch` in this function instead of a mutation. All other requests will be paused while `getAuth` returns, so you won't get multiple auth failures

Reviewer: ```suggestion
use `fetch` in this function instead of a mutation. All other requests will be paused while `getAuth` returns, so we never have to handle multiple auth failures
```

at the same time.

### Configuring `willAuthError`

> `willAuthError` is an optional parameter and is run _before_ a network request is made. You can use it to trigger the logic in
> `getAuth` without the need to get a graphql error first. For example, you could use this indicate that the auth will fail because the JWT is invalid:

Reviewer: ```suggestion
`willAuthError` is an optional parameter and is run _before_ a network request is made. We can use it to trigger the logic in
`getAuth` without the need to send a request and get a GraphQL Error back. For example, we can use this to predict that the authentication will fail because our JWT is invalid already:
```

```js
const willAuthError = ({ authState }) => {
  if (!authState || /* JWT is expired */) return true;
  return false;
}
```

## Handling Logout with the Error Exchange

> It is also possible to handle your logout logic in the error exchange instead of in the auth exchange. To do this, you'll need to add the
> `errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:

Reviewer: ```suggestion
We can also handle authentication errors in an `errorExchange` instead of the `authExchange`. To do this, we'll need to add the
`errorExchange` to the exchanges array, _before_ the `authExchange`. The order is very important here:
```

```js
import { createClient, dedupExchange, cacheExchange, fetchExchange, errorExchange } from 'urql';
import { authExchange } from '@urql/exchange-auth';

const client = createClient({
  url: '/graphql',
  exchanges: [
    dedupExchange,
    cacheExchange,
    errorExchange({

>       onError: ({ error }) => {

Reviewer: ```suggestion
      onError: (error) => {
```

        const isAuthError = error.graphQLErrors.some(
          e => e.extensions?.code === 'FORBIDDEN',
        );

        if (isAuthError) {
          logout();
        }
      }
    }),
    authExchange({
      /* config */
    }),
    fetchExchange,
  ],
});
```

> The `errorExchange` will only receive an auth error when the auth exchange as tried and failed to handle it. This means we have
> either failed to refresh the token, or there is no token refresh functionality. If you get an auth error in the `errorExchange` (as defined in
> the `didAuthError` configuration section above), then you can be confident that it is an auth error, the `authExchange` isn't able to recover
> from it, and the user should be logged out.

Reviewer: ```suggestion
The `errorExchange` will only receive an auth error when the auth exchange has already tried and failed to handle it. This means we have
either failed to refresh the token, or there is no token refresh functionality. If we receive an auth error in the `errorExchange` (as defined in
the `didAuthError` configuration section above), then we can be confident that it is an auth error that the `authExchange` isn't able to recover
from, and the user should be logged out.
```
